### DISCUSSION TITLE
Feature Request: Create mergeable split video files

### DISCUSSION THREAD

**Discussion by Leseratte10** [gopro/labs#276] [General]
The GoPro cameras split recorded video into 4GB or 12GB chunks depending on the camera and the SD card filesystem - but despite searching around the internet and asking around on forums, there seems to be absolutely no way to merge these files together in a way that keeps *all* metadata intact, including the timecode section and the GoPro metadata stream. 

There's a section on the GoPro Labs page - https://gopro.github.io/labs/control/chapters/ - that explains how to merge chapters, but that doesn't work correctly. 

ffmpeg, even in its newest version, can combine the video and audio streams just fine, but A) it modifies the stream metadata and B) it doesn't correctly copy/merge the Timecode and the GoPro metadata stream with all the GPS and Gyroscope data. Neither does the "ReelSteady-Joiner", because it's just a GUI for ffmpeg. The provided udtacopy program from the link above copies the global file metadata, but not the stream metadata, and it also doesn't do anything about the GPMD and TMCD streams. 

This means that I can merge the video files together to view them in normal programs or upload to YouTube or do whatever with them, but I always need to archive / store the original split files as well, just in case I need to access the GPS/Gyro/other GoPro metadata later, because there's no tool that can merge these. 

I see three possible ways this annoying issue could be fixed: 

- A GoPro Labs QR code feature that completely disables chaptering. This should be an option for people who like to have it, but it's also not ideal because one issue with the SD card could ruin a hour-long video if the file gets corrupted.
- Continue splitting the video files, but in a binary-compatible way so one can just do `copy \B GX010009.MP4 + GX020009.MP4 fullvideo.MP4`, or on Linux, `cat GX010009.MP4 GX020009.MP4 > fullvideo.MP4` to get a joined/merged video with all the metadata intact and a proper GPS data stream. So, once the cam hits the 4G/12G limit, just close the file, open a new one, and continue writing, instead of writing a whole new MP4 stream file.
- Release a piece of software similar to the already existing udtacopy that properly joins together chaptered video files, keeping the time code and the GPMD stream completely intact with all its contents and all metadata. Or, ideally, get support for that into the open-source FFMPEG so that that tool can handle losslessly joining GoPro chapters.

Is this something that could potentially be added?

**Comment by dnewman-gpsw** [comment]
I disagree with much of your findings.  Likely the only thing that might need work is timecode, yet ffmpeg should support that.  Otherwise ReelSteady-Joiner works correctly, as does the manual joining instructions which I wrote, as if it didn't, ReelSteedy would not work, it uses all the streaming GPMF data.  I use these tools regularly.  So you should revisit to see what might be going wrong for you.

As for no chaptering, that is harder, as all the indexing is saved upon file close, so more indexes have to be held in memory.  So there are limits.  There is an almost completely untested hack, try oM64BT=32000, this will give you 32GB chapters.  Now this might not work in all capture modes, like 2.7k240 is risky, as high framerate mean more indexes to buffer.  I did a test using 190Mb/s at 24p, and 32GB chapters worked fine.   Low bitrates with high fps, are the biggest memory concern.  I don't use low bitrates, so I'm trying 150Mb/s at 240fps and 64GB chapters now, that will take a while.

**Reply by dnewman-gpsw** [reply]
Follow up, oM64BT=64000, at 2.7kp240 @ 150Mb/s., seems to work, but I run out of battery at 52GB into the capture.  So this is a promising solution.

**Reply by Leseratte10** [reply]
Interesting. I remember testing this with the newest ffmpeg a couple weeks ago and couldn't manage to get a file with the whole metadata stream intact so I tried to reproduce it again after your response and now it seems to merge the GPMD stream ... not sure what went wrong last time, then. Thanks for telling me that that's something that should be working out of the box. The timecode is there and the GoPro metadata seems to be intact, though I didn't test it. 

There's still some stuff that's wrong with newest ffmpeg (it resets the MP4 version and brand and it doesn't seem to copy the stream's "creation_date" and "handler_name" fields for all streams, but the metadata stream is intact, so that must have been fixed by ffmpeg at some point. 

Though, after I created this report this morning I have found another tool to merge MP4 files, [mp4-merge](https://github.com/gyroflow/mp4-merge), which does handle GoPro files even better than ffmpeg and keeps all the internal stream fields intact. It does corrupt the tmcd stream but maybe that bug can be fixed in that tool. 

Thanks for that oM64BT command - I didn't find that in the documentation when searching for a way to disable chaptering. I will give that a try. 

As for "raw" file chaptering (so they can be merged with a simple cat or copy), I'd assume that would have the same issues as a large / infinite chapter size because it needs to keep all the index stuff in RAM?

Thanks for your response and sorry for wasting your time with something that was probably user error to some extent.

**Reply by clement-nardi** [reply]
Thank you for this precious piece of advice!!
Trying to properly merge the gopro video files while keeping all the side streams is a nightmare...
So deactivating the file chaptering "feature" is the perfect solution for me.

I just did a test on a naked GoPro Hero10 with !M64BT=1024000 and !MBITR=200 hoping for a maximum file size of 1TB (my SD card is 256GB), and it worked great, I could record for more than 2 hours, except the file was split in chunks of 100GB!

I wonder if this is a hard maximum value or if the value I sent was truncated to 6 chars (102400 instead of 1024000).
I will do another test with !M64BT=256000 and let you know.

**Reply by dnewman-gpsw** [reply]
Read the rest of the thread.  Chapters are limited to either 64GB or 81 minutes, which every comes first.

**Reply by clement-nardi** [reply]
Well, i tried again and still got 100GB files of 66 minutes.

**Reply by dnewman-gpsw** [reply]
Great, you are in untested territory. I just checked the code and there is no upper limit on the size that Labs applies, it was only tested to 64GB. If you drop the bit-rate, the chapter length will increase to 81 minutes max.  So the two limits seems to be 100GB or 81 minutes.

**Comment by gbuzogany** [comment]
I am having the same issue. 
When opening the chapters the camera recorded, stabilization is fine, but after merging the chapters with `ReelSteady-Joiner`, the footage is really shaky. Stabilization data got messed up somehow. I used the latest version of `ReelSteady-Joiner` and checked the hash of `udtacopy.exe` that `ReelSteady-Joiner` is using to validate that it is using the latest version, and it matched the latest `udtacopy.exe` available in this repo (`labs`).

**Reply by Leseratte10** [reply]
I've recently switched to [mp4merge](https://github.com/gyroflow/mp4-merge) to merge my GoPro files and that seems to be working without any issues and properly preserves both the metadata and the timecode data - give it a try and see if it works for you too.

**Comment by defcomedyjam** [comment]
> Follow up, oM64BT=64000, at 2.7kp240 @ 150Mb/s., seems to work, but I run out of battery at 52GB into the capture. So this is a promising solution.

i think you once replied to me before on the gopro reddit, i tested the 64BT=64000 command and still resulted in split files despite the recording not even reaching the 64000 file size, you said that for some reason the chapter will automatically split at the 81 minute mark, was this resolved? thanks in advance.

**Reply by dnewman-gpsw** [reply]
This is by design, so there are currently no plans to change this, as the work to extend beyond 64GB or 81 minutes, which ever comes first, is significantly more than the work to date. The biggest issue is it would require additional RAM, a precious resource. The indexes for all files recording have to preallocated, so that can't be finitely large, which is one of the reasons for chaptering on most video products.
